extension/arquillian-drone-browserstack-extension/src/main/java/org/arquillian/drone/browserstack/extension/webdriver/BrowserStackDriverFactory.java

```diff
@@ -0,0 +1,130 @@
+ * {@link org.jboss.arquillian.drone.spi.Instantiator} and {@link org.jboss.arquillian.drone.spi.Destructor} for
+ * {@link BrowserStackDriver}.
+ *
+ * @author <a href="mailto:mjobanek@redhat.com">Matous Jobanek</a>
+ */
+public class BrowserStackDriverFactory implements
+    Configurator<BrowserStackDriver, WebDriverConfiguration>,
+    Instantiator<BrowserStackDriver, WebDriverConfiguration>,
+    Destructor<BrowserStackDriver> {
+
+    private static final Logger log = Logger.getLogger(BrowserStackDriverFactory.class.getName());
+
+    @Inject
+    protected Instance<BrowserCapabilitiesRegistry> registryInstance;
+
+    public WebDriverConfiguration createConfiguration(ArquillianDescriptor arquillianDescriptor,
+        DronePoint<BrowserStackDriver> dronePoint) {
+
+        BrowserCapabilitiesRegistry registry = registryInstance.get();
+
+        // first, try to create a BrowserCapabilities object based on Field/Parameter type of @Drone annotated field
+        BrowserCapabilities browser = registry.getEntryFor(READABLE_NAME);
+
+        WebDriverConfiguration configuration = new WebDriverConfiguration(browser).configure(arquillianDescriptor,
+                                                                                             dronePoint.getQualifier());
+
+        return configuration;
+    }
+
+    public void destroyInstance(BrowserStackDriver browserStackDriver) {
+        browserStackDriver.quit();
+    }
+
+    public BrowserStackDriver createInstance(WebDriverConfiguration configuration) {
+        try {
+            Capabilities capabilities = configuration.getCapabilities();
+
+            String url = (String) capabilities.getCapability(URL);
+            String accessKey = null;
+            if (isEmpty(url)) {
+                String username = (String) capabilities.getCapability(USERNAME);
+                accessKey = (String) capabilities.getCapability(ACCESS_KEY);
+                if (isEmpty(accessKey)) {
+                    accessKey = (String) capabilities.getCapability("automate.key");
+                }
+                if (isEmpty(username) || isEmpty(accessKey)) {
+                    log.severe(
+                        "You have to specify either an username and an access.key or the whole url in your arquillian descriptor");
+                    return null;
+                } else {
+                    url = "http://" + username + ":" + accessKey + "@hub.browserstack.com/wd/hub";
+                }
+            }
+
+            boolean isSetBrowserStackLocal = capabilities.is(BROWSERSTACK_LOCAL);
+            boolean isSetBrowserStackLocalManaged = capabilities.is(BROWSERSTACK_LOCAL_MANAGED);
+
+            if (isSetBrowserStackLocal && isSetBrowserStackLocalManaged && (!isEmpty(accessKey) || !isEmpty(url))) {
+                if (isEmpty(accessKey)) {
+                    accessKey = url.substring(url.lastIndexOf(":") + 1, url.indexOf("@"));
```

Comment:
you could potentially do some matcher for url. you expect that it has some format which does not need to be true every time. if that url would be of wrong format, you would throw ex or log it.